### PR TITLE
Use pj for agfj (and others)

### PR DESCRIPTION
**Pull request by blenk92**
As requested in PR https://github.com/radare/radare2/pull/13051 a port of agfj (and other commands) to pj.

The change is quite large, so you maybe want to make sure that i didn't mix up the types of  the variables that are put into the json

### CODE REVIEW

File: libr/util/pj.c [hunk 1]
```diff
@@ -194,6 +202,14 @@ R_API PJ *pj_n(PJ *j, ut64 n) {
 	return j;
 }
 
+R_API PJ *pj_sn(PJ *j, st64 n) {
```
Comment (radare) [inline libr/util/pj.c]: why is this required? the comma should be automatically a dded if the parent element is an array or object. and the raw thing is pj_kd.. ..

ok so u want to represent st64? pj_sn is a bad name for this (signed number) because all 2 letters apis are KEY, VALUE, so its confusing, it should be a single letter, also the meaning of pj_s is misleading this function name. maybe pj_N instead?

Comment (blenk92) [reply libr/util/pj.c]: i just copy-pastaed this from the things below. I'm not 100% sure why this comma thingy is required.
Yeah naming is an issue, i'll fix that.

I don't get what you mean by "the raw thing is pj_kd" from what i see this is for vars of type double?

File: libr/core/cmd_print.c [hunk 2]
```diff
@@ -881,23 +881,29 @@ static void cmd_pDj(RCore *core, const char *arg) {
 	if (bsize < 0) {
 		bsize = -bsize;
 	}
-	r_cons_print ("[");
+	PJ *pj = pj_new ();
```
Comment (radare) [inline libr/core/cmd_print.c]: add nullchk

File: libr/core/cmd_print.c [hunk 3]
```diff
@@ -3639,8 +3645,10 @@ static void disasm_recursive(RCore *core, ut64 addr, int count, char type_print)
 	RAnalOp aop = {0};
 	int ret;
 	ut8 buf[128];
+	PJ *pj = NULL;
 	if (type_print == 'j') {
-		r_cons_print ("[");
+		pj = pj_new ();
```
Comment (radare) [inline libr/core/cmd_print.c]: add nullchk

File: libr/core/cmd_print.c [hunk 4]
```diff
@@ -3776,8 +3784,8 @@ static void func_walk_blocks(RCore *core, RAnalFunction *f, char input, char typ
 	}
 	r_list_sort (f->bbs, (RListComparator) bbcmp);
 	if (input == 'j' && b) {
-		r_cons_print ("[");
-		bool isFirst = true;
+		pj = pj_new ();
```
Comment (radare) [inline libr/core/cmd_print.c]: add nullchk

File: libr/core/cmd_print.c [hunk 5]
```diff
@@ -4802,15 +4802,15 @@ static int cmd_print(void *data, const char *input) {
 					const char *orig_bb_middle = r_config_get (core->config, "asm.bb.middle");
 					r_config_set_i (core->config, "asm.bb.middle", false);
 					cont_size = tmp_get_contsize (f);
-					r_cons_printf ("{");
-					r_cons_printf ("\"name\":\"%s\"", f->name);
-					r_cons_printf (",\"size\":%d", fcn_size);
-					r_cons_printf (",\"addr\":%"PFMT64u, f->addr);
-					r_cons_printf (",\"ops\":[");
+					pj = pj_new ();
```
Comment (radare) [inline libr/core/cmd_print.c]: add nullchk

File: libr/core/canal.c [hunk 6]
```diff
@@ -1154,26 +1154,28 @@ static int core_anal_graph_nodes(RCore *core, RAnalFunction *fcn, int opts) {
...
+		pj_ki (pj, "nlocals",
 			r_anal_var_count (core->anal, fcn, 'r', 0) +
 			r_anal_var_count (core->anal, fcn, 's', 0) +
 			r_anal_var_count (core->anal, fcn, 'b', 0));
-		r_cons_printf (",\"size\":%d", r_anal_fcn_size (fcn));
-		r_cons_printf (",\"stack\":%d", fcn->maxstack);
-		r_cons_printf (",\"type\":\"%s\"", r_anal_fcn_type_tostring (fcn->type));
-		//r_cons_printf (",\"cc\":%d", fcn->call); // TODO: calling convention
+		pj_kn (pj, "size",  r_anal_fcn_size (fcn));
+		pj_ki (pj, "stack", fcn->maxstack);
```
Comment (radare) [inline libr/core/canal.c]: i dont see why maxstack should be negative, this is probably a task to discuss in a different pr.

you are missing some casts to (st64) because r_anal_fcn_size returns ut32

File: libr/core/canal.c [hunk 7]
```diff
@@ -1154,26 +1154,28 @@ static int core_anal_graph_nodes(RCore *core, RAnalFunction *fcn, int opts) {
...
-		r_cons_printf ("{\"name\":\"%s\"", r_str_get (fcn_name_escaped));
+		pj_o (pj);
+		pj_ks (pj, "name", r_str_get (fcn_name_escaped));
 		free (fcn_name_escaped);
-		r_cons_printf (",\"offset\":%"PFMT64d, fcn->addr);
-		r_cons_printf (",\"ninstr\":%"PFMT64d, (ut64)fcn->ninstr);
-		r_cons_printf (",\"nargs\":%d",
+		pj_kn (pj, "offset", fcn->addr);
+		pj_ki (pj, "ninstr", fcn->ninstr);
+		pj_ki (pj, "nargs",
```
Comment (radare) [inline libr/core/canal.c]: i dont see why number of args can be negative :?

File: libr/core/canal.c [hunk 8]
```diff
@@ -1154,26 +1154,28 @@ static int core_anal_graph_nodes(RCore *core, RAnalFunction *fcn, int opts) {
...
 			r_anal_var_count (core->anal, fcn, 'r', 1) +
 			r_anal_var_count (core->anal, fcn, 's', 1) +
 			r_anal_var_count (core->anal, fcn, 'b', 1));
-		r_cons_printf (",\"nlocals\":%d",
+		pj_ki (pj, "nlocals",
```
Comment (radare) [inline libr/core/canal.c]: same for nlocals. we should discuss this in a different pr

File: libr/core/disasm.c [hunk 9]
```diff
@@ -5426,15 +5426,15 @@ R_API int r_core_print_disasm_instructions(RCore *core, int nb_bytes, int nb_opc
...
+R_API int r_core_print_disasm_json(RCore *core, ut64 addr, ut8 *buf, int nb_bytes, int nb_opcodes, PJ *pj) {
 	RAsmOp asmop;
 	RDisasmState *ds;
 	RAnalFunction *f;
 	int i, j, k, ret, line;
 	ut64 old_offset = core->offset;
 	ut64 at;
 	int dis_opcodes = 0;
-	//r_cons_printf ("[");
+	//pj_a (pj);
```
Comment (ret2libc) [inline libr/core/disasm.c]: just remove please

File: libr/core/canal.c (hunk 7 above)
Comment (blenk92) [reply libr/core/canal.c]: From a logical perspective you are completely right. However, i was usually looking at the type of the variables that are added to the json output. For instance `r_anal_var_count()` returns int anyway, so from my perspective it was a better approach to simply use that type.

**Comment by radare** [comment]
Every refactor give us the opportunity to improve the codebase and discuss stuff to make r2 better. So that pj refactoring is also another excuse to find those inconsistencies and fix them
